src/algebra/free_non_unital_non_assoc_algebra.lean

```diff
@@ -0,0 +1,101 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.free
+import algebra.monoid_algebra
+
+/-!
+# Free algebras
+
+Given a semiring `R` and a type `X`, we construct the free non-unital, non-associative algebra on
+`X` with coefficients in `R`, together with its universal property. The construction is valuable
+because it can be used to build free algebras with more structure, e.g., free Lie algebras.
+
+## Main definitions
+
+  * `free_non_unital_non_assoc_algebra`
+  * `free_non_unital_non_assoc_algebra.lift`
+  * `free_non_unital_non_assoc_algebra.of`
+
+## Implementation details
+
+We construct the free algebra as the magma algebra, with coefficients in `R`, of the free magma on
+`X`. However we regard this as an implementation detail and thus deliberately omit the lemmas
+`of_apply` and `lift_apply`, and we mark `free_non_unital_non_assoc_algebra` and `lift` as
+irreducible once we have established the universal property.
+
+## Tags
+
+free algebra, non-unital, non-associative, free magma, magma algebra, universal property,
+forgetful functor, adjoint functor
+-/
```

Comment:
I think it would be good to reference `free_algebra` somewhere here in case the reader wants that instead.

Reply:
Good point. Done.